[zhonghan721] iP

zhonghan721 opened the pull request:
# Zean

> "Time isn't the main thing. It's the only thing." - Miles Davis ([source](https://blog.rescuetime.com/time-management-quotes/))

Zean chatbot helps you keep track of all the things that you have to do, so you don't have to do so.
It is:
- text-based
- _easy_ to use
- ~~CHEAP~~ **FREE** to use

To get started,
1. download it from [here](https://github.com/zhonghan721/ip/releases)
2. double click it (Alternatively, run the command `java -jar zean.jar` in the same folder as the jar file)
3. add your tasks
4. enjoy the convenience 💯 

Features included:
- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)

#

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public static void main(String[] args) {
    new Zean("./data/zean.txt").run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

zhonghan721 committed:
Add Zean.txt and Ui.txt

zhonghan721 committed:
Complete Echo functionality

zhonghan721 committed:
Add Level-2 requirements

zhonghan721 committed:
Add Level-3 functionalities

zhonghan721 committed:
Add Level-4 functionalities

zhonghan721 committed:
Add automated testing test cases

zhonghan721 committed:
Add Level-5 handling of errors functionality

zhonghan721 committed:
Update error checking

zhonghan721 committed:
Add Level-6 delete functionality

zhonghan721 committed:
Add javadocs

zhonghan721 committed:
Delete Duke.java

zhonghan721 committed:
Add read and write ability

The chatbot was unable to store data from a session.

The chatbot reads from the file if present and populates the Array List.
If the file is not present, it creates the file.
Every changes made to the tasks are written to the file.

zhonghan721 committed:
Merge branch 'branch-Level-7'

zhonghan721 committed:
Change deadline Strings to LocalDate

zhonghan721 committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Change deadline Strings to LocalDate

zhonghan721 committed:
Add TaskList.java

There are several methods in Storage class that should be in TaskList class.

Creating the TaskList class improves the OOP aspects of the code.

zhonghan721 committed:
Merge branch 'branch-A-MoreOOP'

* branch-A-MoreOOP:
  Add TaskList.java

zhonghan721 committed:
Organise classes into packages

As more classes are created, it is difficult to trace and read the code.

Relevant classes are grouped into packages to enhance readability.

zhonghan721 committed:
Merge branch 'branch-A-Packages'

* branch-A-Packages:
  Organise classes into packages

zhonghan721 committed:
Merge branch 'master' into add-gradle-support

* master:
  Organise classes into packages
  Add TaskList.java
  Change deadline Strings to LocalDate
  Add read and write ability
  Delete Duke.java
  Add javadocs
  Add Level-6 delete functionality
  Update error checking
  Add Level-5 handling of errors functionality
  Add automated testing test cases
  Add Level-4 functionalities
  Add Level-3 functionalities
  Add Level-2 requirements
  Complete Echo functionality
  Add Zean.txt and Ui.txt

zhonghan721 committed:
Update gradle to work with current chatbot

Gradle support was previously work for only duke.

All necessary changes were made for it to work with zean.
Configured gradle checkstyle to work.

zhonghan721 committed:
Merge branch 'add-gradle-support'

* add-gradle-support:
  Update gradle to work with current chatbot
  Bump gradle and lib version
  Add Gradle support

zhonghan721 committed:
Add JUnit tests

zhonghan721 committed:
Merge branch 'branch-A-JUnit'

* branch-A-JUnit:
  Add JUnit tests

zhonghan721 committed:
build.gradle: Change jar file name to zean

zhonghan721 committed:
Add JavaDoc comments

zhonghan721 committed:
new feature: Add ability to search for task

zhonghan721 committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  new feature: Add ability to search for task

zhonghan721 committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add JavaDoc comments

zhonghan721 committed:
Update code to comply with coding standard

zhonghan721 committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Update code to comply with coding standard

zhonghan721 committed:
Merge branch 'master' into branch-A-MoreOOP

* master:
  Update code to comply with coding standard
  new feature: Add ability to search for task
  Add JavaDoc comments
  build.gradle: Change jar file name to zean
  Add JUnit tests
  Update gradle to work with current chatbot
  Organise classes into packages
  Bump gradle and lib version
  Add Gradle support

zhonghan721 committed:
Update Ui class to handle most of the printing of output

zhonghan721 committed:
Merge branch 'branch-A-MoreOOP'

* branch-A-MoreOOP:
  Update Ui class to handle most of the printing of output

zhonghan721 committed:
Merge branch 'master' into branch-A-JUnit

* master:
  Update Ui class to handle most of the printing of output
  Update code to comply with coding standard
  new feature: Add ability to search for task
  Add JavaDoc comments
  build.gradle: Change jar file name to zean

zhonghan721 committed:
TaskListTest: Add more test cases

zhonghan721 committed:
Merge branch 'branch-A-JUnit'

* branch-A-JUnit:
  TaskListTest: Add more test cases

zhonghan721 committed:
JavaDoc: Update incomplete javadoc

zhonghan721 committed:
Implement GUI

zhonghan721 committed:
Implement GUI for Zean chatbot

zhonghan721 committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Implement GUI for Zean chatbot
  Implement GUI

zhonghan721 committed:
Update code to resolve warnings from checkStyle

zhonghan721 committed:
Merge branch 'branch-A-CheckStyle'

* branch-A-CheckStyle:
  Update code to resolve warnings from checkStyle

elhy1999 commented on src/main/java/zean/Parser.java:
```diff
@@ -0,0 +1,176 @@
+     */
+    public static String parse(String input, TaskList tasks) throws DukeException {
+        if (input.isBlank()) {
+            throw new DukeException("OOPS!!! You forgot to type in anything!");
```
I like how you provided different error messages for handling different possible cases when the user provides invalid inputs 👍

elhy1999 commented on src/main/java/zean/Parser.java:
```diff
@@ -0,0 +1,176 @@
+                throw new DukeException("OOPS!!! You forgot to specify the starting date."
+                        + "\nUse \"/from\" to do so.");
+            }
+            if (indexOfTo == -1) {
+                throw new DukeException("OOPS!!! You forgot to specify the ending date."
+                        + "\nUse \"/to\" to do so.");
+            }
+
+            description = input.substring(5, indexOfFrom);
+            if (description.isBlank()) {
+                throw new DukeException("Hmm, the description of an event cannot be empty :(");
+            }
+            String from = input.substring(indexOfFrom + 5, indexOfTo).strip();
+
+            if (from.isBlank()) {
+                throw new DukeException("OOPS!!! You forgot to specify the starting date.");
+            }
+
+            String to = input.substring(indexOfTo + 3).strip();
+            if (to.isBlank()) {
+                throw new DukeException("OOPS!!! You forgot to specify the ending date.");
+            }
+
+            if (!DATE_PATTERN.matcher(from).matches() || !DATE_PATTERN.matcher(to).matches()) {
+                throw new DukeException("Hmm, I don't understand the date. "
+                        + "Use this format: YYYY-MM-DD");
+            }
+            output = tasks.add(description, from, to);
+            break;
+        case "find":
+            description = input.substring(4);
+            System.out.println(description);
+            if (description.isBlank()) {
+                throw new DukeException("Please provide the keyword for me to search.");
+            }
+            output = tasks.find(description);
+            break;
+        default:
+            throw new DukeException("OOPS!!! I'm sorry, but I don't understand what that means :-(");
+        }
```
I like that you followed to switch-case indentation standard for our class 😄

elhy1999 commented on src/main/java/zean/Storage.java:
```diff
@@ -0,0 +1,112 @@
+     *
+     * @param task The task to be written to the file.
+     */
+    public void addToDisk(Task task) {
```
I also like how you have phrased all your methods as verbs, as we've learnt!

elhy1999 commented on src/main/java/zean/exception/DukeException.java:
```diff
@@ -0,0 +1,14 @@
+ *
+ * @author Zhong Han
+ */
+public class DukeException extends RuntimeException {
```
Should this exception be renamed to ZeanException instead? 🤔

elhy1999 reviewed (commented):
LGTM! Just a few nits to fix. I like how almost all your code follows the coding standards, which made it very easy to read 😄

PohSayKeong commented on src/main/java/zean/Parser.java:
```diff
@@ -0,0 +1,176 @@
+            }
+
+            description = input.substring(5, indexOfFrom);
+            if (description.isBlank()) {
+                throw new DukeException("Hmm, the description of an event cannot be empty :(");
+            }
+            String from = input.substring(indexOfFrom + 5, indexOfTo).strip();
+
+            if (from.isBlank()) {
+                throw new DukeException("OOPS!!! You forgot to specify the starting date.");
+            }
+
+            String to = input.substring(indexOfTo + 3).strip();
+            if (to.isBlank()) {
+                throw new DukeException("OOPS!!! You forgot to specify the ending date.");
+            }
```
specific error handling which is great

PohSayKeong commented on src/main/java/zean/Parser.java:
```diff
@@ -0,0 +1,176 @@
+     * @return The output to be printed on the console.
+     * @throws DukeException An exception related to the chatbot.
+     */
+    public static String parse(String input, TaskList tasks) throws DukeException {
```
this method is a little long, you might want to consider moving some of the logic into their own classes

PohSayKeong commented on src/main/java/zean/Parser.java:
```diff
@@ -0,0 +1,176 @@
+     * @param text The formatted text read from the file.
+     * @return The number of tasks added successfully to the ArrayList.
+     */
+    public static int parseToTask(ArrayList<Task> arrList, String text) {
```
instead of adding an `ArrayList` directly, you might consider passing in a `TaskList` instead.

PohSayKeong commented on src/main/java/zean/Storage.java:
```diff
@@ -0,0 +1,112 @@
+ */
+public class Storage {
+
+    private File f;
```
do try to avoid single letter variable names if possible

PohSayKeong commented on src/main/java/zean/Storage.java:
```diff
@@ -0,0 +1,112 @@
+     * Empty constructor for storage.
+     * For the purpose of testing.
+     */
+    public Storage() {
```
if I'm not wrong if the constructor is empty you can choose to omit it, do double check on this thanks!

PohSayKeong commented on src/main/java/zean/Storage.java:
```diff
@@ -0,0 +1,112 @@
+        this.filePath = filePath;
+
+        File dir = new File("./data");
+        if (!dir.exists()) {
+            if (!dir.mkdir()) {
```
maybe you want to combine these two into a single if statement

PohSayKeong commented on src/main/java/zean/Storage.java:
```diff
@@ -0,0 +1,112 @@
+        } catch (FileNotFoundException e) {
+            throw new DukeException("OOPS! The file cannot be created.");
+        } finally {
+            if (sc != null) {
+                sc.close();
+            }
```
would it be ok if this goes into the try clause?

PohSayKeong commented on src/main/java/zean/task/Deadline.java:
```diff
@@ -0,0 +1,76 @@
+     * @param description The description of the deadline task.
+     * @param by The due date of the deadline task.
+     */
+    public Deadline(String description, String by) throws DukeException {
+        super(description.strip());
+        try {
+            this.by = LocalDate.parse(by.strip());
+        } catch (DateTimeParseException e) {
+            throw new DukeException("The date is invalid!");
+        }
+    }
+
+    /**
+     * Constructor for the deadline task.
+     *
+     * @param bool The completion status of the deadline task.
+     * @param description The description of the deadline task.
+     * @param by The due date of the deadline task in the format YYYY-MM-DD.
+     */
+    public Deadline(String bool, String description, String by) {
+        super(description.strip());
+        this.by = LocalDate.parse(by.strip());
+        if (Integer.parseInt(bool) == 1) {
+            this.isDone = true;
+        }
+    }
```
the logic for these two constructors are somewhat similar, are there way to combine them?

PohSayKeong reviewed (commented):
Looks good overall and you have also written quite comprehensive tests. Keep up the good work!

zhonghan721 committed:
Refactor code

There are several methods that are too long, resulting in redability
issue.

Extracting them out and refactoring improves the code quality, making
it easier to debug.

zhonghan721 committed:
Add java assertions

There are certain assumptions made in some methods.
In cases where the assumptions do not hold, unintended behaviour may
happen.

Using java assertions to verify and handle those assumptions allows for
more graceful handling of the situation.

Let's use java assertions to check for those assumptions moving forward.

zhonghan721 committed:
Merge pull request #1 from zhonghan721/branch-A-CodeQuality

Refactor code

zhonghan721 committed:
Merge branch 'master' into branch-A-Assertions

* master:
  Refactor code

# Conflicts:
#	src/main/java/zean/TaskList.java
#	src/main/java/zean/task/Deadline.java
#	src/main/java/zean/task/Event.java
#	src/main/java/zean/task/Todo.java

zhonghan721 committed:
Merge pull request #2 from zhonghan721/branch-A-Assertions

Add java assertions

zhonghan721 committed:
Merge branch 'master' of https://github.com/zhonghan721/ip

* 'master' of https://github.com/zhonghan721/ip:
  Add java assertions

zhonghan721 committed:
Implement update feature

Currently, there is no way for users to update desciption/dates realated
to the tasks.

Adding such a feature will make it more convinient for users to update
their tasks.

zhonghan721 committed:
Merge branch 'branch-C-Update'

* branch-C-Update:
  Implement update feature

zhonghan721 committed:
Set up continuous integration

zhonghan721 committed:
Update code to comply with checkstyle

zhonghan721 committed:
Update GUI

The speech dialogue boxes are of a single color, causing a lack of
differentiation between Zean and the user.

Changing the color of the user dialogue box makes it easier for the
user. Adding another bright color to indicate error message alerts
the users.

zhonghan721 committed:
Merge branch 'branch-A-BetterGui'

* branch-A-BetterGui:
  Update GUI

zhonghan721 committed:
Add user guide and screenshot

zhonghan721 committed:
Merge branch 'branch-A-UserGuide'

* branch-A-UserGuide:
  Add user guide and screenshot

zhonghan721 committed:
Update user guide
